controllers/storagecluster/consumer.go

```diff
@@ -0,0 +1,189 @@
+package storagecluster
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	providerClient "github.com/red-hat-storage/ocs-operator/services/provider/client"
+)
+
+var (
+	// externalOCSResources will hold the ExternalResources for storageclusters
+	// ExternalResources can be accessible using the UID of an storagecluster
+	// getExternalConfigFromProvider func sets value for this once and use same again
+	externalOCSResources map[types.UID][]ExternalResource
+)
+
+type ocsConsumer struct{}
+
+// ensureCreated performs onboarding and update capacity
+func (o *ocsConsumer) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !isExternalOCSProvider(instance) {
+		return nil
+	}
+
+	if instance.Status.ExternalStorage.ConsumerID == "" {
+		return onBoardConsumer(r, instance)
+	}
+
+	if instance.Spec.ExternalStorage.RequestedCapacity != nil &&
+		!instance.Status.ExternalStorage.GrantedCapacity.Equal(*instance.Spec.ExternalStorage.RequestedCapacity) {
+		return upscaleOrDownscaleConsumer(r, instance)
+	}
+
+	return nil
+}
+
+// ensureDeleted performs offboarding
+func (o *ocsConsumer) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !isExternalOCSProvider(instance) {
+		return nil
+	}
+
+	return offBoardConsumer(r, instance)
+}
+
+// onBoardConsumer makes an API call to the provider cluster for onboarding
+func onBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	grpcConnection, consumerClient, err := getConsumerClient(instance)
+	if err != nil {
+		return err
+	}
+	defer grpcConnection.Close()
+
+	// TODO: think about naming convention
+	response, err := consumerClient.OnboardConsumer("my-ocs", getExternalStorageRequestedCapacity(instance))
+	err = r.handleOnboardConsumerErrors(instance, err)
+	if err != nil {
+		return err
+	}
+
+	if response.StorageConsumerUUID == "" || response.GrantedCapacity == "" {
+		return fmt.Errorf("response is empty")
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return nil
+}
+
+// offBoardConsumer makes an API call to the provider cluster for offboarding
+func offBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	grpcConnection, consumerClient, err := getConsumerClient(instance)
```

Comment:
how about creating client  and closing the connection only once inside the `ensureCreated` func.   The client can be passed in as an argument or can be added to StorageClusterReconciler Struct.

Reply:
It is not just this `ensureCreated` where these API calls are used, these are used at other places also, in that case, I need to write the logic at other `ensureCreated` too.

I don't want to make it part of the reconciler, because we do not need these calls very frequently and do not want to get the client every time.

Reply:
IMO, the current code makes it impossible/difficult to unit test the onBoard, offboard and updatedCapacity functionality. The client should be passed in as a dependency via `ocsConsumer` struct.